src/nvim/diff.c

```diff
@@ -1733,25 +1743,713 @@ void diff_clear(tabpage_T *tp)
   diff_T *next_p;
   for (p = tp->tp_first_diff; p != NULL; p = next_p) {
     next_p = p->df_next;
+    if (!p->df_redraw) {
```

Comment:
I'd suggest removing the conditions at all. Either p->df_comparisonlines is NULL and xfree() will do nothing, or it's not NULL and free() will release the memory.